src/characters/porter.ts

```diff
@@ -0,0 +1,148 @@
+  result: {
+    ursulas: UrsulaResponse[];
+  };
+  version: string;
+}
+
+interface PostRetrieveCFragsRequest {
+  treasure_map: Base64EncodedBytes;
+  retrieval_kits: Base64EncodedBytes[];
+  alice_verifying_key: HexEncodedBytes;
+  bob_encrypting_key: HexEncodedBytes;
+  bob_verifying_key: HexEncodedBytes;
+  policy_encrypting_key: HexEncodedBytes;
+}
+
+interface PostRetrieveCFragsResult {
+  result: {
+    retrieval_results: {
+      cfrags: {
+        [address: string]: string;
+      };
+    }[];
+  };
+  version: string;
+}
+
+export class Porter {
+  private readonly porterUri: string;
+
+  constructor(porterUri: string) {
+    this.porterUri = porterUri;
+  }
+
+  public async getUrsulas(
+    quantity: number,
+    durationPeriods: number,
+    excludeUrsulas?: ChecksumAddress[],
+    includeUrsulas?: ChecksumAddress[]
+  ): Promise<Ursula[]> {
+    const params: GetUrsulasRequest = {
+      quantity,
+      duration_periods: durationPeriods,
+      exclude_ursulas: excludeUrsulas,
+      include_ursulas: includeUrsulas,
+    };
+    const resp: AxiosResponse<GetUrsulasResponse> = await axios.get(
+      `${this.porterUri}/get_ursulas`,
+      {
+        params,
+        paramsSerializer: (params) => {
+          return qs.stringify(params, { arrayFormat: 'comma' });
+        },
+        headers: {
+          'Access-Control-Allow-Origin': this.porterUri,
+        },
+      }
+    );
+    return resp.data.result.ursulas.map((u: UrsulaResponse) => ({
+      checksumAddress: u.checksum_address,
+      uri: u.uri,
+      encryptingKey: u.encrypting_key,
+    }));
+  }
+
+  public revoke(revocations: RevocationRequest[]): RevocationResponse {
+    throw new Error('Method not implemented.');
+  }
+
+  public async retrieveCFrags(
+    treasureMap: TreasureMap,
+    retrievalKits: RetrievalKit[],
+    aliceVerifyingKey: PublicKey,
+    bobEncryptingKey: PublicKey,
+    bobVerifyingKey: PublicKey,
+    policyEncryptingKey: PublicKey
+  ): Promise<RetrievalResult[]> {
+    const data: PostRetrieveCFragsRequest = {
+      treasure_map: toBase64(treasureMap.toBytes()),
+      retrieval_kits: retrievalKits.map((rk) => toBase64(rk.toBytes())),
+      alice_verifying_key: toHexString(aliceVerifyingKey.toBytes()),
+      bob_encrypting_key: toHexString(bobEncryptingKey.toBytes()),
+      bob_verifying_key: toHexString(bobVerifyingKey.toBytes()),
+      policy_encrypting_key: toHexString(policyEncryptingKey.toBytes()),
+    };
+    const resp: AxiosResponse<PostRetrieveCFragsResult> = await axios.post(
+      `${this.porterUri}/retrieve_cfrags`,
+      data,
+      {
+        headers: {
+          'Access-Control-Allow-Origin': this.porterUri,
+        },
+      }
+    );
+    return resp.data.result.retrieval_results
+      .map((result) => result.cfrags)
+      .map((cFrags) => {
+        const parsed = Object.keys(cFrags).map((address) => [
+          address,
+          VerifiedCapsuleFrag.fromVerifiedBytes(fromBase64(cFrags[address])),
```

Comment:
This skips the verification process, correct? Does this mean that the cfrag bytes being returned by Porter are implicitly trusted as being correct? Should we be verifying the cfrag returned and not accepting it as correct without actual verification?

Reply:
That's a good point, these should definitely be verified.